Name: Tarun Tanmay Class: MBATech CE , Sem 5
Roll Number: N049 

In [2]:
#Experiment 8 
#Data Cleaning
import numpy as np #used for handling numbers
import pandas as pd #used for handling datasets
from sklearn.impute import SimpleImputer #used for missing data
from sklearn.preprocessing import LabelEncoder #used to convert categorical variable into corresponding number

In [5]:
df=pd.read_csv('data_preprocess_1.csv')
df

Region   Age   Income Online shopper
0   India  49.0  86400.0             No
1  Brazil  32.0  57600.0            Yes
2     USA  35.0  64800.0             No
3  Brazil  43.0  73200.0             No
4     USA  45.0      NaN            Yes
5   India  40.0  69600.0            Yes
6  Brazil   NaN  62400.0             No
7   India  53.0  94800.0            Yes
8     USA  55.0  99600.0             No
9   India  42.0  80400.0            Yes

In [11]:
x=df.iloc[:,:-1].values
y=df.iloc[:,-1].values
y

array(['No', 'Yes', 'No', 'No', 'Yes', 'Yes', 'No', 'Yes', 'No', 'Yes'],
      dtype=object)

In [16]:
#Handling missing values 
impu=SimpleImputer(missing_values=np.nan,strategy='mean')
impu=impu.fit(x[:,1:])
x[:,1:]=impu.transform(x[:,1:])
x

array([['India', 49.0, 86400.0],
       ['Brazil', 32.0, 57600.0],
       ['USA', 35.0, 64800.0],
       ['Brazil', 43.0, 73200.0],
       ['USA', 45.0, 76533.33333333333],
       ['India', 40.0, 69600.0],
       ['Brazil', 43.77777777777778, 62400.0],
       ['India', 53.0, 94800.0],
       ['USA', 55.0, 99600.0],
       ['India', 42.0, 80400.0]], dtype=object)

mean of age is = 43.77 , mean of Income is = 76533.33

In [19]:
#Handling Categorical type of data
label_y=LabelEncoder()
y=label_y.fit_transform(y)
y

array([0, 1, 0, 0, 1, 1, 0, 1, 0, 1])

No is replaced by 0, Yes is replaced by 1

In [22]:
z=df.iloc[:,0]
label_z=LabelEncoder()
z=label_z.fit_transform(z)
z

array([1, 0, 2, 0, 2, 1, 0, 1, 2, 1])

z represents the region column which is converted into numbers 0,1,2 for mapping them with age and Income

In [24]:
df.iloc[:,0]=z
df.iloc[:,1]=x[:,1]
df.iloc[:,2]=x[:,2]
df.iloc[:,3]=y
df


Region      Age   Income  Online shopper
0       1       49    86400               0
1       0       32    57600               1
2       2       35    64800               0
3       0       43    73200               0
4       2       45  76533.3               1
5       1       40    69600               1
6       0  43.7778    62400               0
7       1       53    94800               1
8       2       55    99600               0
9       1       42    80400               1

The Data Set is clean now 

Conclusion:
1) If data set has missing values for multiple columns having numbers then missing values can be replaced either with mean or median or any option given in SimpleImputer funtion.
2) It is always better to convert or map category type data values to corresponding representative numbers.
3) A column has three discrete values then corresponding numbers will be 0, 1, and 2. 
3) It is advisable to preprocess the data set by replacing missing values by meaningful values, and mapping categorical values into corresponding numbers before giving them to any classifier. 

In [45]:
#Naive Bayes Classifier 
import pandas as pd 
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB

In [47]:
data=pd.read_csv('Naive_bayes_1.csv')
data

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0              1         0       3  ...   7.2500   NaN         S
1              2         1       1  ...  71.2833   C85         C
2              3         1       3  ...   7.9250   NaN         S
3              4         1       1  ...  53.1000  C123         S
4              5         0       3  ...   8.0500   NaN         S
..           ...       ...     ...  ...      ...   ...       ...
886          887         0       2  ...  13.0000   NaN         S
887          888         1       1  ...  30.0000   B42         S
888          889         0       3  ...  23.4500   NaN         S
889          890         1       1  ...  30.0000  C148         C
890          891         0       3  ...   7.7500   NaN         Q

[891 rows x 12 columns]

In [51]:
#convert category to numeric
data['Sex_cleaned']=np.where(data['Sex']=='male',0,1)
data['Embarked_cleaned']=np.where(data['Embarked']=='S',0,np.where(data['Embarked']=='C',1,np.where(data['Embarked']=='Q',2,3)))
#Clean data with Nan Values 
data=data[['Survived','Pclass','Sex_cleaned','Age','SibSp','Parch','Fare','Embarked_cleaned']]

In [52]:
data=data.dropna(axis=0,how='any')
data.shape

(714, 8)

In [64]:
x_train, x_test=train_test_split(data,test_size=0.2,random_state=4)
gnb=GaussianNB()
features=['Pclass','Sex_cleaned','Age','SibSp','Parch','Fare','Embarked_cleaned']
gnb.fit(x_train[features].values,x_train['Survived'])
y_pred=gnb.predict(x_test[features])

In [65]:
#Print Values
numb_test=x_test.shape[0]
numb_mislabled=(x_test['Survived']!=y_pred).sum()
accuracy=(1-(numb_mislabled/numb_test))*100
print('Performance {: 05.2f}%'.format(accuracy))

Performance  79.02%


Conclusion: 
1) Before applying Naive bayes classifier to the titanic data set, all the rows with Nan is removed and category type of columns are mapped to their representative numbers.
2) We created a model for the cleaned data set and applied Naive Bayes Classification for the test size = 50% and we got accuracy as 76%.
3) The accuracy is 79% by increasing train size to 80% from 50%. 
